### ISSUE TITLE
Lambda response is always base64 encoded

### ISSUE THREAD

**Issue by robertzlatarski** [serverless-nextjs/serverless-next.js#316]
Seems like the body of requests made to the `/api `endpoint are always base64 encoded, when using the Serverless component. We are currently making a switch from the plugin to the component and since we are running GraphQL on our `/api` path, this is not something we want.

Looking at the code where this potentially happens: https://github.com/danielcondemarin/serverless-next.js/blob/84f3840e1c49238c5a9cb6e144947aa56357f3f9/packages/serverless-nextjs-component/next-aws-cloudfront.js#L147

Why would we want to encode the body of the response? Shouldn't it be opt-in and definitely not happening if the response is a valid JSON?

**Comment by mattdell** [comment]
Yeah we're wondering about this as well. Can we return it without base64 encoding?

**Comment by mattdell** [comment]
Any thoughts on this @danielcondemarin?

**Comment by mattdell** [comment]
Just to add @danielcondemarin, I'm getting this as a response from one of my `/api` endpoints in NextJS.

```
H4sIAAAAAAAAA6tWSi0qyi8qVrKKrlbKTS0uTkxPVbJSCkgtKs4sLklNCSxNLar0yy9xyy/NS1HSUUqtKEnNK87MzwPqqFZKzk8Bq3YNCvYMDnF1iQ8MdQ2KjPfzD4l38w/1c1GqrY2t5QIA0Ck+i2QAAAA=
```

It should be a JSON response, but I can't seem to decode it to anything resembling JSON to save my life. I've tried decoding it from base64 and from gzipped base64 using `zlib` (looking at next-aws-cloudfront.js) and it's just jibberish.

Can the API just return JSON?

**Comment by danielcondemarin** [comment]
I can't replicate this. I've just put together a simple api endpoint that works,

![image](https://user-images.githubusercontent.com/1122442/88090828-0f3e0380-cb86-11ea-9e36-10d239e57ddd.png)

```js
export default (req, res) => {
  res.statusCode = 200;
  res.setHeader("Content-Type", "application/json");
  res.end(JSON.stringify({ name: "John Doe" }));
};
```

Could you give more details about your api endpoint, code etc. that may help track this down?

**Comment by mattdell** [comment]
Thanks very much for looking into this @danielcondemarin. You're right, a simple lambda response returns just fine.

We are using Apollo Server to connect to GraphQL. We've done some more investigation and this seems to be around using persisted queries in Apollo Server. If we disable persisted queries then it works. If we enable persisted queries then it doesn't work. We can't disable persisted queries, though, as that will remove a lot of caching on our servers.

We've noticed that using persisted queries uses a GET request with a querystring, but non persisted queries uses a POST with data. I'm currently investigating what difference that might make.

We have also noticed that we're falling into the `gz` ternary: https://github.com/serverless-nextjs/serverless-next.js/blob/12da1de31855b68b9addef801ec21dffd3202a21/packages/lambda-at-edge-compat/next-aws-cloudfront.js#L234-L236

I've console logged each branch and here are the responses: 

```
// zlib
H4sIAAAAAAAAA3WPTQvCMAyG7/6MgLfBxGNBRLx6G55HXLNZ1g9p04mM/XfbOWQIHt8nD0neESQyghghBvIVI8eQkwqnyHeyrBpkkiBa1IEKsGgIBEABN+d6g74PZxctg9gVEHAgWRH65k4rjHIgzytABpX+bNHK9vNB7TplEyuxabJYJuDsMXp92O7bbKYc+deIvFI8dSow+SwZKi09E1zshZFUnGBd8+tBS5nrt/ll/mb6O4dpmjZvRIaYhzMBAAA=

// regularBuffer:
eyJkYXRhIjp7InVzZXJTdGF0dXMiOnsiaXNBdXRoZW50aWNhdGVkIjpmYWxzZSwibmFtZSI6IiIsImJvb2ttYXJrc0NvdW50IjowLCJzYXZlZFNlYXJjaGVzQ291bnQiOjAsImFkdmVydHNDb3VudCI6MCwiZW1haWwiOiIiLCJsaW5rcyI6eyJsb2dpbiI6Ii9hY2NvdW50L2xvZ29uP3VybD0lMmYiLCJsb2dvdXQiOiIvYWNjb3VudC9sb2dvdXQ/dXJsPSUyZiIsInJlZ2lzdGVyIjoiL21lL25ldyIsImFjY291bnQiOiIvbWUvZWRpdCIsIl9fdHlwZW5hbWUiOiJVc2VyU3RhdHVzTGlua3MifSwiX190eXBlbmFtZSI6IlVzZXJTdGF0dXMifX19Cg==
```

The major difference here is that the `responsePromise` is returning the zlib response which gets decoded as base64 as jibberish

```
uMÂ0ïþ·ÁÄcAD¼zG\³YÖiÓýwÛ9dß'IÞ$2!ò#Ç
§Èw²¬d ZÔ
°h@7çz¾g-Øp YúæN+r Ï+@þlÑÊöóAí:e+±i²X&àì1z}ØîÛl¦ù×¼R<u*0ù,*-=\ìT`]óëAKë·ùeþfú;i6oD3
```

But the regular buffer decodes to valid base64
```
{"data":{"userStatus":{"isAuthenticated":false,"name":"","bookmarksCount":0,"savedSearchesCount":0,"advertsCount":0,"email":"","links":{"login":"/account/logon?url=%2f","logout":"/account/logout?url=%2f","register":"/me/new","account":"/me/edit","__typename":"UserStatusLinks"},"__typename":"UserStatus"}}}
```

Note this happens on both the POST and GET methods. I thought I had cracked it until I realized the POST (which works fine) also returns the gzipped jibberish version. Unless I console logged the wrong thing. 🤷🏻‍♂️

Sorry that is mostly just a brain dump of where I am. I'm continuing to research, but if you have any ideas they would be greatly appreciated.

**Comment by mattdell** [comment]
@danielcondemarin one thing you might be able to help explain is what's going on in this `responsePromise` method.

Taking the simple example above it enters the method with text as a json string parameter

```
text: '{"name":"John Doe"}'
```

Which then gets written with `res.write` and then it continues on.

Tthe `response.body` is a Buffer but then gets base64 encoded and gzipped into 

```
response.body = 'H4sIAAAAAAAAA6tWykvMTVWyUvLKz8hTcMlPVaoFAARYjGwTAAAA' // base64 + gzip of buffer
```

So the response will be written in plain text but then the promise gets resolved somwhere with `resolve(response);`. Where is this happening?

I wonder if whatever is calling the promise is erroring or falling into a conditional branch somewhere that I can't see.

When I console log `text` on my code it's fine:

```
{"data":{"userStatus":{"isAuthenticated":false,"name":"","bookmarksCount":0,"savedSearchesCount":0,"advertsCount":0,"email":"","links":{"login":"/account/logon?url=%2f","logout":"/account/logout?url=%2f","register":"/me/new","account":"/me/edit","__typename":"UserStatusLinks"},"__typename":"UserStatus"}}}
```

But I don't understand why that doesn't get written but whatever is sent to this promise resolver as `response.body` is what ultimately ends up getting written.

```
H4sIAAAAAAAAA6tWSi0qyi8qVrKKrlbKTS0uTkxPVbJSCkgtKs4sLklNCSxNLar0yy9xyy/NS1HSUUqtKEnNK87MzwPqqFZKzk8Bq3YNCvYMDnF1iQ8MdQ2KjPfzD4l38w/1c1GqrY2t5QIA0Ck+i2QAAAA=
```

halp

**Comment by danielcondemarin** [comment]
@mattdell Could you DM me [here](https://twitter.com/dcondemarin). I'll invite you to our slack workspace.

**Comment by mattdell** [comment]
For some reason this is working if I return `text` rather than `respnse.body` in the `responsePromise`

```
response.body = gz
  ? zlib.gzipSync(text).toString("base64")
  : Buffer.from(text).toString("base64");
```

Compare with https://github.com/serverless-nextjs/serverless-next.js/blob/12da1de31855b68b9addef801ec21dffd3202a21/packages/lambda-at-edge-compat/next-aws-cloudfront.js#L223-L237

Could `text` be used instead of `response.body`? It seems to work for my situation coming back from Apollo Server / GraphQL as well as a simple Hello World endpoint. 

Also, I don't see what difference it makes, but adding the extra condition `&& text` to the if statement above also fixes the problem. 🤷🏻‍♂️

```
if (response.body && text) {
  response.bodyEncoding = "base64";
  response.body = gz
    ? zlib.gzipSync(response.body).toString("base64")
    : Buffer.from(response.body).toString("base64");
}
```

**Comment by mattdell** [comment]
Confirmed fixed with `1.16.0-alpha.1`

Fine work @danielcondemarin! 🎉
